pkg/security/podsecuritypolicy/provider.go

```diff
@@ -449,3 +452,13 @@ func validateRuntimeClassName(actual *string, validNames []string) field.ErrorLi
 	}
 	return field.ErrorList{field.Invalid(field.NewPath("spec", "runtimeClassName"), *actual, "")}
 }
+
+func allowsVolumeType(allowedVolumes sets.String, fsType policy.FSType, volume api.Volume) bool {
+	if allowedVolumes.Has(string(fsType)) {
+		return true
+	}
+	if allowedVolumes.Has(string(policy.Secret)) && fsType == policy.Projected && utilfeature.DefaultFeatureGate.Enabled(features.BoundServiceAccountTokenVolume) && reflect.DeepEqual(*volume.VolumeSource.Projected, *serviceaccount.TokenVolumeSource) {
```

Comment:
DeepEqual is more fragile than we want, and we really just care about the information exposed via the volumes

Make a method in `pkg/security/podsecuritypolicy` that takes an `*api.ProjectedVolumeSource` and returns true if it only contains service-account-token-related sources (paying attention only to the information exposed):

```go
func IsOnlyServiceAccountTokenSources(v *api.ProjectedVolumeSource) bool {
  for _, s := range v.Sources {
    if t := s.ServiceAccountToken; t != nil && (t.Path != "token" || t.Audience != "") {
      return false
    }

    if s.ConfigMap != nil && s.ConfigMap.LocalObjectReference.Name != "kube-root-ca.crt" {
      return false
    }

    if s.DownwardAPI != nil {
      for _, d := range s.DownwardAPI.Items {
        if d.Path != "namespace" || d.FieldRef == nil || || d.FieldRef.APIVersion != "v1" || d.FieldRef.FieldPath != "metadata.namespace" {
          return false
        }
      }
    }
  }
}
```

Then add a unit test that ensures that `IsOnlyServiceAccountTokenSources(serviceaccount.TokenVolumeSource ())` returns true

Reply:
if there is no ServiceAccountToken or ConfigMap or DownwardAPI, it will return true. we would only allow if all three sources match.

Reply:
PSP is interested in protecting access to unauthorized data. Allowing anything that fits within the data a projected service account token volume would expose seems reasonable.

Reply:
done